Feature/issue 3501 graphgl

dmfrey opened the pull request:
<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
BeanDefinitions: Mix instanceSupplier and props

dmfrey committed:
WIP: initial implementation

dmfrey committed:
Merge remote-tracking branch 'origin/main' into feature/issue_3501_graphgl

dmfrey committed:
fix checkstyle issues

dmfrey committed:
fix checkstyle issues

dmfrey committed:
git ignore jenv local version file

dmfrey committed:
git ignore jenv local version file

artembilan reviewed (changes requested):
Just general feedback: Spring GraphQL does depend on Spring Boot: https://github.com/spring-projects/spring-graphql/blob/main/build.gradle#L8.

Even if it looks like `spring-graphql` module is clean, the project by itself has an extra `graphql-spring-boot-starter` module. So, Spring Boot depends on  Spring Integration, and Spring GraphQL on Spring Boot. Therefore we can't use that project in our code base.

It is perfectly fine to advertise Spring GraphQL for target projects, but we would need to find some way how to smoothly use their API...

dmfrey committed:
WIP: make it reactive

dmfrey committed:
WIP: make it reactive

artembilan reviewed (changes requested):
Some review so far to let you progress.

Thank you!

dmfrey committed:
remove references to `.java-version`

dmfrey committed:
remove references to `.java-version`

dmfrey committed:
implement reactive endpoint for GraphQL Query

issue #3501

dmfrey committed:
implement reactive endpoint for GraphQL Query

issue #3501

dmfrey committed:
refactor to handle GraphQL Query and Mutation requests

issue #3501

dmfrey committed:
refactor to handle GraphQL Query and Mutation requests

issue #3501

dmfrey committed:
refactor to handle GraphQL Subscription requests

issue #3501

dmfrey committed:
refactor to handle GraphQL Subscription requests

issue #3501

dmfrey committed:
Merge remote-tracking branch 'origin/main' into feature/issue_3501_graphgl

rstoyanchev reviewed (commented):
In addition to inline comments, it seems there is quite a bit of duplication across the  Query, Mutation, and Subscription tests, and I'm wondering if it couldn't be one test class, with a controller that has all 3 kinds of methods?

dmfrey committed:
refactor to reflect feedback

issue #3501

dmfrey committed:
refactor to reflect feedback

issue #3501

artembilan reviewed (changes requested):
I think this is good so far: just a couple general project usage concerns and some evolutionary ideas.

Also: it is probably won't hurt to start already some doc on the matter. Just initial draft would be enough: some `graphql.adoc` in the `/src/reference/asciidoc` folder and the link in the `index.adoc` and `index-single.adoc`.

Thank you!

dmfrey committed:
Merge branch 'spring-projects:main' into feature/issue_3501_graphgl

dmfrey committed:
implement requested feedback

issue #3501

dmfrey committed:
implement requested feedback

issue #3501

dmfrey committed:
implement expressions to handle for various RequestInput parameters

issue #3501

dmfrey committed:
implement expressions to handle for various RequestInput parameters

issue #3501

dmfrey committed:
implement expressions to handle for various RequestInput parameters

issue #3501

dmfrey committed:
implement expressions to handle for various RequestInput parameters

issue #3501

artembilan committed:
Rework `JpaOutboundGatewayParserTests` to JUnit 5

artembilan committed:
Fix Checkstyle violation in the XML config

dmfrey committed:
implement requested changes

issue #3501

dmfrey committed:
implement requested changes

issue #3501

artembilan reviewed (changes requested):
Some further review after looking closely.

dmfrey committed:
implement requested changes

issue #3501

dmfrey committed:
implement requested changes

issue #3501

dmfrey committed:
rename tests for clarity

issue #3501

dmfrey committed:
rename tests for clarity

issue #3501

dmfrey committed:
implement requested changes

issue #3501

dmfrey committed:
implement requested changes

issue #3501

artembilan committed:
GH-3653: Improve DSL parsing performance (#3654)

* GH-3653: Improve DSL parsing performance

Fixes https://github.com/spring-projects/spring-integration/issues/3653

* Refactor `IntegrationFlowBeanPostProcessor` to perform `beanFactory.getBeansOfType()`
as less as possible
* Implement a `NamedComponent` on the `StandardIntegrationFlow` for better
bean scanning performance in the `IntegrationFlowBeanPostProcessor`

* * Skip `NamedComponent` from bean scanning if `beanName == null`

artembilan committed:
GH-3656: Fix package tangles

Fixes https://github.com/spring-projects/spring-integration/issues/3656

* Move `JavaUtils` from `util` package to the root one to break any possible
tangling to/from other packages.
* Deprecate an existing `JavaUtils` for backward compatibility
* Remove `@IntegrationConverter` from the `JsonNodeWrapperToJsonNodeConverter`
and register it manually in the `ConverterRegistrar`.
The bean registration for the `JsonNodeWrapperToJsonNodeConverter` must be removed
in the next `6.0`
* Remove usage of `IntegrationContextUtils` from the `support` package

artembilan committed:
Add `toReactivePublisher(autoStartOnSubscribe)` (#3657)

* Add `toReactivePublisher(autoStartOnSubscribe)`

The `IntegrationFlowBuilder.toReactivePublisher()` returns a `Publisher<Message<?>>`
which may be subscribed somewhere late in the application logic, e.g.
when WebSocket (or RSocket) subscription is initiated by the external client.
In between application context startup and that subscription moment, the `IntegrationFlow`
must not try to produce messages since there is nothing to consumer them from the
`Publisher<Message<?>>` side.
One of the way is to have a source endpoint not started automatically and control its
lifecycle from the point fo reactive subscription

* Introduce an `IntegrationFlowBuilder.toReactivePublisher(boolean autoStartOnSubscribe)`
to let the framework do a job for an `IntegrationFlow` lifecycle control.
This way end-user doesn't need to know autowire a starting endpoint and use `doOnSubscribe()`
and similar callbacks
* Change `ConsumerEndpointFactoryBean` log message about a `FixedSubscriberChannel`
to `INFO` since an `autoStartup = false` really does not have any effect and there is nothing
for end-user to worry about.
The `IntegrationFlow` knows nothing about each endpoint internals and cannot control which
mark as `autoStartup = false` and which not

* Fix languge in JavaDocs

Co-authored-by: Gary Russell <grussell@vmware.com>

Co-authored-by: Gary Russell <grussell@vmware.com>

artembilan committed:
Start 6.0 version

* Upgrade to Java 17, SF-6.0, Gradle 7.2
* Upgrade to Jakarta dependencies and respective namespaces
* Fix some tests for Java 17 compatibility
* Fix wrong Javadocs
* Add some missed Javadocs
* Fix more `jakarta` namespace
* Fix WS & XML modules to use Jakarta EE
* `--add-opens` in some modules for their reflection-based tests
* Disable Kafka tests which does not work on Windows; see Apache Kafka `3.0.1`
* Upgrade to JUnit `5.8.1`
* Migrate JMS tests to Artemis
* Remove RMI module as it was deprecated before
* Fix `pr-build-workflow.yml` for Java 17
* Fix JavaDocs warnings using `Xdoclint:syntax` per module, not in the top-level `api` task
* Move docs for version `6.0`

mauromol committed:
GH-3652: File gateway: don't try delete no file

Fixes spring-projects/spring-integration#3652

Do not try to delete non-existing file in `FileExistMode.REPLACE` in the `AbstractRemoteFileOutboundGateway`

**Cherry-pick to `5.4.x` & `5.3.x`**

artembilan committed:
Use ReflectionUtils.makeAccessible() - not direct

The `LambdaMessageProcessor` and `MethodInvokingMessageSource`
 call `method.setAccessible(true)` unconditionally.

* Use `ReflectionUtils.makeAccessible()` for some optimization
* Remove `Content is not allowed in prolog` from the `JaxbMarshallingIntegrationTests`
since it is locale dependant and not relevant for the unit test logic

StandCN committed:
Fix typo in the MqttMessageSentEvent Javadoc

artembilan committed:
`DefaultConfiguringBFPP` to `BDRegistryPP`

* Make a `DefaultConfiguringBeanFactoryPostProcessor` as a `BeanDefinitionRegistryPostProcessor`
to let it be processed properly in Spring AOT
* Register `DefaultConfiguringBeanFactoryPostProcessor` as a `BeanDefinition.ROLE_INFRASTRUCTURE`,
so it is removed in Spring AOT after processing and writing target bean definitions

artembilan committed:
Fix unused import in the `IntegrationRegistrar`

Meteorkor committed:
GH-3655: Add automatically delete for Redis Locks

Fixes https://github.com/spring-projects/spring-integration/issues/3655

* support automatically clean up cache
* RedisLockRegistry.capacity desc

artembilan committed:
Fix WebFluxMH for proper response handling

The `WebFluxRequestExecutingMessageHandler` does direct `ClientResponse.create(entity.getStatusCode())`
which comes with a `ExchangeStrategies.withDefaults()`.
Even if end-user configures a `WebClient` properly, the response is created with default strategies.

* Rework `WebFluxRequestExecutingMessageHandler` internal logic to call `ResponseSpec.toEntityFlux(BodyExtractor)`
instead of manual `ClientResponse.create()`
* Add unit test to the `WebFluxRequestExecutingMessageHandlerTests` to ensure that configured `maxInMemorySize`
on the `WebClient` strategies has an effect when response body is bigger than expected size

**Cherry-pick to `5.4.x`**

artembilan committed:
Fix new Sonar smells

artembilan committed:
Revert method references back to lambdas

Turns out Java does cast for the method reference argument before really calling
a delegating method.

* Revert cast method references (`((ApplicationContextAware) this.handler)::setApplicationContext`)
back to lambdas - `context -> ((ApplicationContextAware) this.handler).setApplicationContext(this.applicationContext)`
This way the cast is deferred until the lambda is called.

artembilan committed:
Revert `AnnotationGatewayProxyFactoryBean` logic

The `AnnotationGatewayProxyFactoryBean` can be configured via setters
and via annotation in the provided interface.

* Override `setProxyDefaultMethods()` to `true` only if annotation explicitly
provides `true`

artembilan committed:
Take Docker registry from the DOCKER_REGISTRY_URL

Add a logic into the `TestUtils` to take a Docker registry URL
from the `DOCKER_REGISTRY_URL` ENV variable.
Use the value in the `GenericContainer` for image to pull.
Fallback to an official registry if no value

Meteorkor committed:
Change RedisLockRegistry cache to 100K

1_000_000 of cache entries take too much memory

artembilan committed:
Fix more Sonar smells

Meteorkor committed:
GH-3672: Clean up Jdbc & ZK LockRegistry caches

Fixes https://github.com/spring-projects/spring-integration/issues/3672

* Clean up `JdbcLockRegistry`, `ZookeeperLockRegistry` cache automatically 
* setCapacity(int capacity) to cacheCapacity(int capacity)
* field rename `capacity`to `cacheCapacity`, add static

artembilan committed:
Some docs clean up

artembilan committed:
Some docs clean up for ZK & XML modules

artembilan committed:
Fix Checkstyle violations

* Make version block for dependencies as code-switcher in the XML, ZeroMQ and Zookeeper docs

artembilan committed:
GH-3675: Restore nativeHeaders for WebSocketInChA

Fixes https://github.com/spring-projects/spring-integration/issues/3675

The `WebSocketInboundChannelAdapter` removes a `SimpMessageHeaderAccessor.NATIVE_HEADERS`
from headers used to create an outbound message.
Turns out not all native headers are mapped by the `StompHeaderAccessor`
to the top-level header entries.

**Cherry-pick to `5.4.x` & `5.3.x`**

artembilan committed:
GH-3677: Doc for URL conn customization in FeedCA (#3678)

* GH-3677: Doc for URL conn customization in FeedCA

Fixes https://github.com/spring-projects/spring-integration/issues/3677

If there is need to have a `URLConnection` customized, the `UrlResource`
has to be used instead of plain `URL` injection into the `FeedEntryMessageSource`

* * Add a sample to docs for connection customization

artembilan committed:
Upgrade dependencies; prepare for release

spring-builds committed:
[artifactory-release] Release version 5.5.6

spring-builds committed:
[artifactory-release] Next development version

artembilan committed:
Upgrade To Gradle 7.3

artembilan committed:
Fix checksum for Gradle wrapper

artembilan committed:
Add duplicatesStrategy to prepAsciidocBuild task

artembilan committed:
Upgrade Dokka plugin to 1.5.30 for Gradle 7.3

artembilan committed:
Check for `Disposables` bean before registration

The `MessagingAnnotationPostProcessor` registers a `Disposables` unconditionally.
The `afterPropertiesSet()` maybe called several times, so `IntegrationContextUtils.DISPOSABLES_BEAN_NAME`
registration has to be idempotent.

* Add `if (!this.beanFactory.containsBeanDefinition(IntegrationContextUtils.DISPOSABLES_BEAN_NAME))`
before trying to register `Disposables` bean

artembilan committed:
Remove `Disposables` abstraction

* Remove `Disposables` since it is a package protected therefore could not be used outside
* Remove the logic relevant ot the `Disposables` in the `MessagingAnnotationPostProcessor` logic
* Replace a `registerSingleton()` and `initializeBean()` with the proper `registerBeanDefinition()`
and `getBean()` usage.
This way bean are going to be destroyed properly
* Remove `MessagingAnnotationPostProcessorChannelCreationTests` since its mocking logic is too vague
and really covered with many other real tests.

xak2000 committed:
GH-3683: Always new TX in DefaultLockRepository

Fixes https://github.com/spring-projects/spring-integration/issues/3683

If a transaction is already active while `JdbcLockRegistry` uses
`DefaultLockRepository` to acquire/release a lock, the repository
must execute SQL queries in a separate transaction to prevent
problems with blocking, deadlocking etc.

It also allows to properly follow transaction isolation level that
is set on some methods of `DefaultLockRepository`.

Previously all methods of DefaultLockRepository supported the current
transaction if there are any. Now all methods will always create new
transaction on each call.

* Change tests to be more unit-ish
* Change `@since` for the `DefaultLockRepositoryTests` to `5.3.10`

**Cherry-pick to `5.4.x` & `5.3.x`**

artembilan committed:
Add MessageHandlerMethodFactoryCreatingFactoryBean

Rework the logic around `ManagedList` in the `DefaultConfiguringBeanFactoryPostProcessor`
when it registers a `IntegrationContextUtils.MESSAGE_HANDLER_FACTORY_BEAN_NAME` and
`IntegrationContextUtils.LIST_MESSAGE_HANDLER_FACTORY_BEAN_NAME` into a dedicated
`MessageHandlerMethodFactoryCreatingFactoryBean` with pure Java code instead of
`BeanDefinition` burden

* Deprecate `IntegrationContextUtils.DISPOSABLES_BEAN_NAME` constant for
already removed internal bean
* Align JVM args for Gradle with Java 17 on the `6.0` branch

artembilan committed:
Quote CONDITION whenever necessary for JDBC

SO: https://stackoverflow.com/questions/70286480/how-can-spring-integration-5-5-x-use-mysql-as-message-store

Turns out the condition word is reserved in many SQL DB vendors, e.g.:
https://dev.mysql.com/doc/refman/8.0/en/keywords.html

* Fix SQL scripts for quoting `CONDITION` column name for
those vendors which have it as a reserved word
* Fix `JdbcMessageStore` to have a `CONDITION` quoted to `""` by default and
replaced to "`" for MySQL
* Add `MySqlContainerTest` to test against MySQL Docker container
* Looks like quoted identifiers work well even if they are not reserved words in the SQL vendor

artembilan committed:
GH-3688: Fix WS DSL for proper values propagation

Fixes https://github.com/spring-projects/spring-integration/issues/3688

The `WebServiceTemplate` is populated with some defaults from its ctor.
We should rely on the target default template values as much as possible
and don't override them to `null` if end-user doesn't ask about that explicitly

* Fix `BaseWsOutboundGatewaySpec` extensions to populate values to the target gateway
only if they are not null - therefore provided by end-user
* If end-user wants them explicitly `null`, it is better to do that via
externally configured template.
See overloaded variants for DSL: `Ws.marshallingOutboundGateway(WebServiceTemplate)`
and `Ws.simpleOutboundGateway(WebServiceTemplate)`

**Cherry-pick to `5.4.x` & `5.3.x`**

Meteorkor committed:
GH-3690: Use pub-sub in RedisLockReg for unlocks

Fixes https://github.com/spring-projects/spring-integration/issues/3690

Replace spinLock with pub-sub

rename and add comment

* change remove timing for 'notifyMap'
* use unlink
* remove DUMMY_CALLABLE

* Clean up code style

artembilan committed:
Fix MySqlContainerTest for compatibility

* Upgrade Gradle to `7.3.1`
* Upgrade Log4j to `2.16.0`

artembilan committed:
GH-3694: Resolve circular dep in PublisherAnnBPP (#3695)

* GH-3694: Resolve circular dep in PublisherAnnBPP

Fixes https://github.com/spring-projects/spring-integration/issues/3694

* Move check for `PublisherAnnotationBeanPostProcessor` bean into the
`SmartInitializingSingleton.afterSingletonsInstantiated()` to break a circular dependency revealed
in the `InitializingBean.afterPropertiesSet()`

* * Fix unused import

artembilan committed:
Add distributionSha256Sum prop for Gradle wrapper

artembilan committed:
Fix new Sonar smell

artembilan committed:
Fix UNLINK/DEL logic in the RedisLockRegistry

See review: https://github.com/spring-projects/spring-integration/pull/3691#issuecomment-994124360

artembilan committed:
Fix Checkstyle violation

artembilan committed:
GH-3697: Various lifecycle fixed for MQTT v5 CAs

Fixes https://github.com/spring-projects/spring-integration/issues/3697
SO: https://stackoverflow.com/questions/70374046/spring-integration-mqtt-failed-to-start-app-when-the-network-is-disconnected

* Add `mqttClient.disconnect()` to `Mqttv5PahoMessageDrivenChannelAdapter.doStop()` - the `doStart()` does `connect()`
* Add `Mqttv5PahoMessageDrivenChannelAdapter.destroy()` impl to close `mqttClient`
* Fix `Mqttv5PahoMessageHandler.doStart()` to not re-throw an exception on connection.
Emit an `MqttConnectionFailedEvent` and log error instead
* Fix `Mqttv5PahoMessageHandler.destroy()` to call `mqttClient.close(true)` for better resources clean up
* Improve MQTT v5 components Javadocs and add a reconnect note into `mqtt.adoc`

artembilan committed:
Fix RedisLockRegistry Sonar smells

* Remove redundant `subscribeLock()` method
* Rework `lock()` and `lockInterruptibly()` logic in favor of `while (true) {`
to avoid "empty `while()`" smell

artembilan committed:
Upgrade dependencies including Gradle

* Fix warnings about doc tasks

spring-builds committed:
[artifactory-release] Next development version

artembilan committed:
Improve Kotlin DSL

* Add missed factories, introduced recently into Java DSL
* Add `@IntegrationDsl` - a Kotlin-specific `@DslMarker` annotation
to mark classes (including Java) which are used in Kotlin DSL as a builder pattern.
This way the scope of the builder is honored and IDE does not suggest functions from
higher lambda for builder
* Fix `element-list` reference for Dokka plugin configuration from the respective module

artembilan committed:
GH-3699: Fix typos in the IntegrationFlow javadoc

Fixes https://github.com/spring-projects/spring-integration/issues/3699

shvo123 committed:
GH-3705: Close TcpNioConn.ChannelOutStr.selector

Fixes https://github.com/spring-projects/spring-integration/issues/3705

Closing/destroying `ChannelOutputStream` object does not close the selector therefore it retains redundant pipes/FD that cen be seen using lsof command or ls /proc/

* Close `TcpNioConnection.ChannelOutputStream.selector` in the `ChannelOutputStream`
* Close `TcpNioConnection.ChannelOutputStream` when connection is closed
* Code style clean up

**Cherry-pick to `5.3.x` & `5.4.x`**

artembilan committed:
More Kotlin DSL improvements

Mini256 committed:
Fix variable name typo in the redis.adoc

artembilan committed:
Move IntegrationDsl from .kt to .java

Turns out Eclipse Kotlin plugin is not perfect at the moment,
and it causes some code parsing problems in the editor.

* Move Kotlin annotation to Java to mitigate normal source code flow:
 Java to Kotlin, not opposite

garyrussell committed:
GH-3701: Fix Possible TCP Memory Leak

Resolves https://github.com/spring-projects/spring-integration/issues/3701

Ensure `TcpSender.removeDeadConnection` is always called, for example when
intercepted and closed via `factory.closeConnectionId` or when closed
connections are harvested from the `connections` map.

**Cherry-pick to 5.4.x, 5.3.x**

Meteorkor committed:
GH-3700: Fix Redis lock retry until expiration

Fixes https://github.com/spring-projects/spring-integration/issues/3700

Fixed a case where lock acquisition attempt was abandoned earlier than expected time.
If a vm uses `tryLock()` and another vm unlock it, it will wake up with an unlock message and try to acquire the lock, 
but if it fails(the other vm obtain lock), vm will return false.

artembilan committed:
Fixes according latest upgrades

* Add `--add-opens` to `asciidoctor` Gradle task to avoid compilation warning
* Upgrade to Kotlin `1.6.10` and modify its Gradle task respective
* Add empty impl for new `AnnotationMetadataAdapter.getDeclaredMethods()`
* Parse HTTP methods to new `String methodNames` property of the `RequestMapping`
since `HttpMethod` is not an `enum` anymore.
* Ignore JDBC tests which rely on a map property resolution: the `[]` placeholder
is not handled in the latest SF anymore.
* Fix `WebFluxInboundEndpoint` for deprecations

artembilan committed:
Fix typo in StompInboundChannelAdapter JavaDocs

Fixes https://github.com/spring-projects/spring-integration/issues/3710

artembilan committed:
Upgrade to H2 2.0.206

Turns out something is broken in H2 `Parser` for stored procedures.
So, parameters are now doubled which is definitely not expected.

* Disable those H2 tests where parametrized stored procedures are used

artembilan committed:
Upgrade Dokka plugin to `1.6.10`

The version `1.5.30` is missed somehow on Gradle Plugin Portal

artembilan committed:
Add checkClasspathForConflicts Gradle task

* The new `checkClasspathForConflicts` will fail if modules has dependencies
containing the same classes.
One of the dependency must be excluded to avoid target classpath pollution
and possible conflicts
* Make the task working as `onlyIf { isCI }`
* Fix warning for illegal access in the `asciidoctorPdf` task add `--add-opens`
to its JVM
* Fix `DefaultHeaderChannelRegistry` for proper `MessageChannelWrapper` access
and change it to `record`

artembilan committed:
Fix `ControlBusTests` for DefaultHeaderChannelReg

We cannot use a `DFA` to set value to the `record` property:
it is marked as `trustedFinal` and cannot be modified via reflection.
Therefore, rework `ControlBusTests` to do other logic than reflection.
The reflection in the unit test is not a justification to stay away
from the `record` structure

artembilan committed:
Upgrade dependencies; prepare for release

artembilan committed:
Fix typo in the `build.gradle`

spring-builds committed:
[artifactory-release] Release version 5.5.8

spring-builds committed:
[artifactory-release] Next development version

artembilan committed:
Upgrade dependencies; prepare for release

* Upgrade to Gradle `7.3.3`
* Upgrade to the latest milestones for releasing
* Upgrade to Kryo-5.2.1 and update source code according breaking changes
* Upgrade to H2-2.1.210 and re-enable respective stored procedure tests

spring-builds committed:
[artifactory-release] Release version 6.0.0-M1

spring-builds committed:
[artifactory-release] Next development version

garyrussell committed:
GH-3713: TCP: Fix Intercepted Sender List

Resolves https://github.com/spring-projects/spring-integration/issues/3713

https://github.com/spring-projects/spring-integration/issues/3326 added support
for multiple `TcpSenders`. However, when connections are intercepted, the sender
list was not properly chained through the interceptors.

- override `registerSenders` and properly capture the real senders in the last
  interceptor and intermediate interceptors
- this ensures that `addNewConnection` is called on each interceptor
- when removing dead connections, use the connection sender list insted of the
  factory's raw sender list; detect if the connection is an interceptor and
  call its remove method instead.

**cherry-pick to 5.5.x**

artembilan committed:
Fix various sporadic test failures

* Increase a `default-delay="1000"` for `GatewayInterfaceTests`
to test the late reply scenario
* Use a `ConcurrentHashMap` for `JdbcMessageStore.queryCache`
to avoid a possible `ConcurrentModificationException`

artembilan committed:
GH-3711: Fix HTTP handler for content type header

Fixes https://github.com/spring-projects/spring-integration/issues/3711

The `contentType` header may come with parameter in its media type.

* Fix `AbstractHttpRequestExecutingMessageHandler` to use `equalsTypeAndSubtype()`
ignoring params
* Some other code clean up in the `AbstractHttpRequestExecutingMessageHandler`
* Ensure in the `HttpRequestExecutingMessageHandlerTests.simpleStringKeyStringValueFormData()`
that provided `contentType` header is handled properly
* Fix `HttpProxyScenarioTests.testHttpMultipartProxyScenario()` for mislead multi-part form
handling

**Cherry-pick to `5.5.x`**

garyrussell committed:
GH-3713: Fix Race In Test

garyrussell committed:
Fix @Poller Javadoc

* Fix @Poller Javadoc - `receiveTimeout` was added as a property in 5.1.
* More javadoc polishing.

**cherry-pick to 5.5.x**

artembilan committed:
Clean up of RMI in the docs

garyrussell committed:
Add TCP Test Diagnostics

Meteorkor committed:
GH-3716 Fix wait for init redisMsgListenContainer

Fixes https://github.com/spring-projects/spring-integration/issues/3716

If the `redisMessageListenerContainer` is starting, waiting for it to complete without doing `subscribeUnlock()`

* Introduce `isRunningRedisMessageListenerContainer` state since the `running` in the `RedisMessageListenerContainer` is set in the beginning of the `start()` misleading on the concurrent calls to the `RedisLockRegistry`

**Cherry-pick to `5.5.x`**

almogtavor committed:
Fix sentence error in the overview.adoc

Fix a mistake in a sentence about EIP components and `BeanPostProcessor`

making committed:
Fix applySequence example in the doc

artembilan committed:
Fix race conditions in JdbcLockRegDifClientTests

The default `TTL` in the `DefaultLockRepository` is 10 seconds which mislead us
when other client instance has just 100 millis.
So, not-adjusted client spins for expired lock wastefully leading to the
synchronization barriers to fail

**Cherry-pick to `5.4.x`**

artembilan committed:
Fix some typos and language in docs

* Mostly driver by suggestions from IDEA language plugin
* Add Maven-Gradle code switch for dependency in the modules

migroskub committed:
Add example to Spring Integration with Reactive Streams

* Add outbound channel adapter example

* Add an example for the `CustomReactiveMessageHandler` usage

* Emphasize that the examples are for Reactive Streams

* Fix a missing letter that caused the text to get inside the code block

* Emphasize that the first example is an event driven inbound channel adapter

* Fix a redundant question mark from the example code

artembilan reviewed (changes requested):
Hi @dmfrey !

Any chances to rebase your PR to the latest `main`?
Revise dependency.
And may be look a fresh look into the code 😄 

I wish to have this released in the next week's milestone.

Thank you!

dmfrey commented:
@artem I'll pull in the updates and give it a push.

On Tue, Feb 15, 2022 at 1:17 PM Artem Bilan ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> Hi @dmfrey <https://github.com/dmfrey> !
>
> Any chances to rebase your PR to the latest main?
> Revise dependency.
> And may be look a fresh look into the code 😄
>
> I wish to have this released in the next week's milestone.
>
> Thank you!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spring-projects/spring-integration/pull/3644#pullrequestreview-883438443>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHHFTEDZG34AYXCVTWLMYLU3KKCDANCNFSM5F5PONDA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
My Google Profile <http://www.google.com/profiles/dmfrey>

dmfrey committed:
make current with latest changes in 'spring-graphql'

issue #3501

dmfrey committed:
make current with latest changes in 'spring-graphql'

issue #3501

artembilan commented on spring-integration-graphql/src/main/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandler.java:
```diff
@@ -100,6 +107,23 @@ public void setVariablesExpression(Expression variablesExpression) {
+	 * Set a Execution Id for GraphQL Query to execute.
+	 * @param executionId the executionId to use.
+	 */
+	public void setExecutionId(String executionId) {
```
Well, maybe exactly this `executionId` should be evaluated really against request message.
Therefore an expression. 

I don't know details of this property, so what is your take ?

Thanks

rstoyanchev replied:
`executionId` should be unique per request. There is a GraphQL Java `ExecutionIdProvider` that could be set on `graphql.GraphQL` and by default there is one that creates a random UUID. However it's also possible to set the `executionId` directly on `ExecutionInput` for each request. The Spring MVC and WebFlux transports set it to an HTTP request id, while the WebSocket transports set it to the subscription id of the GraphQL over WebSocket protocol. It can also be overridden from a `WebInterceptor` by using `RequestInput#executionId`.

All in all, a `setExecutionId(String)` on `GraphQlMessageHandler`, which I think is a singleton, doesn't look right indeed. It would have to be a generator callback of some kind and then there is the question of what input do you need to determine it. This could be dropped for now until there is more clarity on the actual need.

artembilan replied:
Thanks, Rossen!

So, yeah, my guess that it has to be unique, per-message is correct. 
Therefore we need to go with a SpEL expression over here where incoming `Message` is going to be root evaluation object. 
Similar to what we do now with that `evaluateOperationNameExpression()`.
I believe by default it can be null to let the target GraphQL engine to generate such an id for us.

artembilan reviewed (changes requested):
Doesn't look like you have rebased this to the latest `main`...

Can you also, please, turn this into a real PR so we can run GH action against it and be sure that build is green?

artembilan commented on spring-integration-graphql/src/main/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandler.java:
```diff
@@ -0,0 +1,170 @@
+
+	private Expression variablesExpression = new SupplierExpression<>(() -> Collections.emptyMap());
+
+	private Locale locale = null;
```
Must `@Nullable` as well.
No need in the `= null`. Because it is `null` by default anyway.

artembilan commented on spring-integration-graphql/src/main/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandler.java:
```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2021-2022 the original author or authors.
```
Just `2022`.
These classes haven't made it to production yet 😄

artembilan commented on build.gradle:
```diff
@@ -100,6 +100,7 @@ ext {
 	soapVersion = '1.4.0'
 	springAmqpVersion = project.hasProperty('springAmqpVersion') ? project.springAmqpVersion : '2.3.11'
 	springDataVersion = project.hasProperty('springDataVersion') ? project.springDataVersion : '2021.0.6'
+	springGraphqlVersion = '1.0.0-SNAPSHOT'
```
I want to see here a milestone since we are going to release next week.

artembilan commented on spring-integration-graphql/src/main/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandler.java:
```diff
@@ -0,0 +1,170 @@
+import java.util.Locale;
+import java.util.Map;
+
+import javax.annotation.Nullable;
```
Must be `org.springframework.lang.Nullable`

artembilan commented on spring-integration-graphql/src/test/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandlerTests.java:
```diff
@@ -0,0 +1,481 @@
+						.hasMessageContaining(
+								"'queryExpression' must not be null"));
+
+	}
```
I see 3 tests with the same content.
I don't think we need many of them since `operationExpression` logic is the same for any GraphQL operation setting on the message handler.

artembilan commented on spring-integration-graphql/src/test/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandlerTests.java:
```diff
@@ -0,0 +1,481 @@
+		Locale locale = Locale.getDefault();
+		String executionId = this.idGenerator.generateId().toString();
+		this.graphQlMessageHandler.setLocale(locale);
+		this.graphQlMessageHandler.setExecutionId(executionId);
```
Why do we need to set these options if we are going to send the whole `RequestInput` anyway?
Shouldn't it be verified in some other unit test when we  determine those `RequestInput` props from request message really?

dmfrey replied:
It's really only relevant when the payload is not a `RequestInput` and is more generic. I've removed them from this test and are placing them in the correct ones.

artembilan commented:
@almogtavor ,

What does it mean?
There is no concept like "reactive app". There could be some services in your app which are reactive, e.g. WebFlux for HTTP. Or this GraphQL message handler.
But that's already fully up to you how you use them in your app.
Reactive is a programming paradigm, not an app design.

So, please, elaborate your concern with more specific info.

Thanks

dmfrey committed:
implement requested changes

issue #3501

dmfrey committed:
Merge remote-tracking branch 'origin/feature/issue_3501_graphgl' into feature/issue_3501_graphgl

# Conflicts:
#	build.gradle
#	spring-integration-graphql/src/main/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandler.java
#	spring-integration-graphql/src/test/java/org/springframework/integration/graphql/outbound/GraphQlMessageHandlerTests.java

dmfrey commented:
Closing for https://github.com/spring-projects/spring-integration/pull/3729
